### ISSUE TITLE
[Question] configuration for reproducing TransE on FB15K

### ISSUE THREAD

**Issue by daizh** [facebookresearch/PyTorch-BigGraph#31]
I have already reproduced the result of `ComplEx` on FB15k, following the example config. (thank you for your works on this great framework, it's really good to use and saves a lot of effort.) 

But when I try to implement `TransE` model as the document suggests, changing `operator` to `translation` and `comparator` to `cos` and `loss_fn` to `ranking` , the result(run by `examples/fb15k.py`) is much lower than expected. 

Playing with `eval_fraction` option, it seems to me that 50 epoch is not enough for the model to coverage. How many epoch is enough?
Or maybe my `margins` param should not be default? Then what should I set `margins` to ?
My configuration details are listed below. 

Thanks you !

## Observed Results
this is filtered result under `examples/fb15k.py`
```
Stats: pos_rank:  97.5809 , mrr:  0.245797 , r1:  0.0258841 , r10:  0.651868 , r50:  0.858958 , auc:  0.983393 , count:  59071
```
this is no-filtered result
```
Stats for edge path 1 / 1: pos_rank:  249.24 , mrr:  0.152477 , r1:  0.0205685 , r10:  0.438252 , r50:  0.701723 , auc:  0.982343 , count:  59071
```
by the way , the final epoch(50th) loss is `12.0212`

## Expected Results

In the paper(Table 2), the TransE result is `MRR=0.594, Hit@10=0.785`

### my config file
```
def get_torchbiggraph_config():

    config = dict(
        entity_path=entity_base,

        num_epochs=50,

        entities={
            'all': {'num_partitions': 1},
        },

        relations=[{
            'name': 'all_edges',
            'lhs': 'all',
            'rhs': 'all',
            'operator': 'translation',
        }],
        dynamic_relations=True,

        edge_paths=[],

        checkpoint_path='model/fb15k_noeval',

        dimension=400,
        global_emb=False,
        comparator='cos',
        loss_fn='ranking',
        lr=0.1,
        num_uniform_negs=1000,

        eval_fraction=0,  # to reproduce results, we need to use all training data
    )

    return config
```
### my conda enviroment:
```
# packages in environment at /home/work/anaconda2/envs/biggraph:
#
# Name                    Version                   Build  Channel
attrs                     19.1.0                   py36_1
blas                      1.0                         mkl
ca-certificates           2019.1.23                     0
certifi                   2019.3.9                 py36_0
cffi                      1.12.2           py36h2e261b9_1
cudatoolkit               9.0                  h13b8566_0
cudnn                     7.3.1                 cuda9.0_0
h5py                      2.9.0            py36h7918eee_0
hdf5                      1.10.4               hb1b8bf9_0
intel-openmp              2019.3                      199
libedit                   3.1.20181209         hc058e9b_0
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 8.2.0                hdf63c60_1
libgfortran-ng            7.3.0                hdf63c60_0
libstdcxx-ng              8.2.0                hdf63c60_1
mkl                       2019.3                      199
mkl_fft                   1.0.10           py36ha843d7b_0
mkl_random                1.0.2            py36hd81dba3_0
ncurses                   6.1                  he6710b0_1
ninja                     1.9.0            py36hfd86e86_0
numpy                     1.16.2           py36h7e9f1db_0
numpy-base                1.16.2           py36hde5b4d6_0
openssl                   1.1.1b               h7b6447c_1
pip                       19.0.3                   py36_0
pycparser                 2.19                     py36_0
python                    3.6.8                h0371630_0
pytorch                   1.0.1           cuda90py36h8b0c50b_0
readline                  7.0                  h7b6447c_5
setuptools                40.8.0                   py36_0
six                       1.12.0                   py36_0
sqlite                    3.27.2               h7b6447c_0
tk                        8.6.8                hbc83047_0
tqdm                      4.31.1                   py36_1
wheel                     0.33.1                   py36_0
xz                        5.2.4                h14c3975_4
zlib                      1.2.11               h7b6447c_3
```

**Comment by daizh** [comment]
It seems I used softmax loss for TransE, it should be ranking loss instead.

**Comment by lerks** [comment]
Sorry for the late reply.

I don't remember who ran that experiment and with what configuration. I have a config for FB15k with TransE lying around which achieves the following stats:
```
pos_rank:  72.3282 , mrr:  0.475054 , r1:  0.318422 , r10:  0.750859 , r50:  0.885883 , auc:  0.973625 , count:  59071
```
which are better than the ones you were getting but still worse than the ones in the paper. I'm posting this config here below, but I'll keep looking for the config we ran those experiments with.
```
entity_base = "data/FB15k"


def get_torchbiggraph_config():

    config = dict(
        entity_path=entity_base,

        num_epochs=200,

        entities={
            'all': {'num_partitions': 1},
        },

        relations=[{
            'name': 'all_edges',
            'lhs': 'all',
            'rhs': 'all',
            'operator': 'translation',
        }],
        dynamic_relations=True,

        edge_paths=[],

        checkpoint_path='model/fb15k',

        dimension=100,
        global_emb=False,
        max_norm=1,
        comparator='dot',
        loss_fn='ranking',
        margin=0.2,
        lr=0.1,
        num_uniform_negs=0,
        num_batch_negs=100,

        eval_fraction=0,  # to reproduce results, we need to use all training data
    )

    return config
```

**Comment by ledw** [comment]
We got numbers that's better than those we reported in the paper for TransE:

    Stats: pos_rank:  58.1476 , mrr:  0.668699 , r1:  0.559327 , r10:  0.843527 , r50:  0.916956 , auc:  0.974048 , count:  59071

Here's config:

    def get_torchbiggraph_config():

        config = dict(
            entity_path=entity_base,

            num_epochs=50,

            entities={
                'all': {'num_partitions': 1},
            },

            relations=[{
                'name': 'all_edges',
                'lhs': 'all',
                'rhs': 'all',
                'operator': 'translation',
            }],
            dynamic_relations=True,
            edge_paths=[],
            checkpoint_path='model/fb15k',

            dimension=400,
            global_emb=False,
            comparator='dot',
            loss_fn='softmax',
            lr=0.1,
            num_uniform_negs=1000,

            eval_fraction=0,  # to reproduce results, we need to use all training data
        )

        return config

**Comment by lerks** [comment]
The differences between the two configs that were posted above are:
- dim: 100 -> 400
- loss_fn: ranking -> softmax
- max_norm: 1 -> none
- num_batch_negs: 100 -> 50
- num_uniform_negs: 0 -> 1000

@ledw ran some ablation experiments, starting from the latest config and resetting one option at a time to be equal to the earlier config, to see which changes have the biggest impact on the MRR:
- dim: 400 -> 100
  ```
  pos_rank:  43.8787 , mrr:  0.64014 , r1:  0.517911 , r10:  0.836764 , r50:  0.920003 , auc:  0.987667
  ```

- loss: softmax -> ranking
  ```
  pos_rank:  98.5577 , mrr:  0.309776 , r1:  0.179623 , r10:  0.562789 , r50:  0.796812 , auc:  0.964517
  ```

- loss: softmax -> ranking and use cosine distance
  ```
  pos_rank:  388.779 , mrr:  0.195723 , r1:  0.133602 , r10:  0.313572 , r50:  0.478204 , auc:  0.976249
  ```

- max_norm: none -> 1
  ```
  pos_rank:  42.7176 , mrr:  0.661254 , r1:  0.541069 , r10:  0.849622 , r50:  0.925336 , auc:  0.982936
  ```

- negatives: 50 -> 100 batch, 1000 -> 0 uniform
  ```
  pos_rank:  44.1177 , mrr:  0.624536 , r1:  0.496927 , r10:  0.831161 , r50:  0.918937 , auc:  0.984603
  ```

**Comment by lerks** [comment]
I guess the main takeaway here is that it's the loss function that really brings big gains. Second to that, but quite far behind, there's the negative sampling. The two might correlate with each other, of course. I'm closing this as I think we've answered the initial request and did some additional research. Feel free to reopen if there's more to talk about.
